### PR TITLE
mkrelease: Update upgrade tests as part of finishing the release

### PR DESCRIPTION
**Pull request by quodlibetor**
`latest_version` tests should always be converted to their actually released
version as soon as that version is released.

In the spirit of eventually getting the release process fully automated, or
at least not regressing the amount of work we need to do as part of it,
this automates the update.

This code is potentially pretty fragile since it expects exact strings
inside the modified yaml file. That said, expecting exactly those strings
and the current structure of the file does allow the script to at least
correctly error if the file changes in a suprising way. Additionally, this
is sort of a best-effort change with a low blast radius/risk profile. If the
modifications fail catastrophically it should be caught either in code
review or by the tests to merge the changes, and in that case we are
basically no worse off than having to modify the file by hand.

### CODE REVIEW

**Review by philip-stoev — APPROVED** [review]
This is correct, thanks for thinking about this and automating it.

The mzcompose.yml file also needs to be modified to include the new version -- it would be cumbersome but not impossible to script it, so can we instead add a manual checklist item to some checklist ?

**Comment by quodlibetor** [comment]
> it would be cumbersome but not impossible to script it

I added this to the script.

File: misc/python/materialize/cli/mkrelease.py [hunk 1]
```diff
@@ -274,6 +276,67 @@ def update_versions_list(released_version: Version) -> None:
...
+            previous = line.split(":")[1].strip()
+        if "TESTS:" in line:
+            tests = line.split(":")[1].strip()
+
+        if "upgrade-from-latest:" in line:
+            insert_workflow = i - 1
+            workflow = f"""
+  upgrade-from-{workflow_version}:
+    env:
+      PREVIOUS_VERSION: {previous}
```
Comment (philip-stoev) [inline misc/python/materialize/cli/mkrelease.py:318]: `PREVIOUS_VERSION` is `released_version`, that is, the version you are upgrading from. For example:

```
  upgrade-from-0_6_1: <- this version
    env:
      PREVIOUS_VERSION: v0.6.1<- is the same as this one
```

With your code, what is generated is:

```
  upgrade-from-0_8_4: 
    env:
      PREVIOUS_VERSION: v0.8.3 <- this should have been 0.8.4 here
```

File: misc/python/materialize/cli/mkrelease.py [hunk 2]
```diff
@@ -274,6 +276,67 @@ def update_versions_list(released_version: Version) -> None:
...
+      TESTS: any_version|{tests}|{version}
```
Comment (philip-stoev) [inline misc/python/materialize/cli/mkrelease.py:319]: `tests` already uncludes `any_version`, so there is no need to mention it explicitly. Your code generates:

```
TESTS: any_version|any_version|v0.6.1|....
```

File: misc/python/materialize/cli/mkrelease.py [hunk 3]
```diff
@@ -274,6 +276,67 @@ def update_versions_list(released_version: Version) -> None:
...
+    steps:
+      - step: workflow
+        workflow: test-upgrade
+"""
+
+        if insert_workflow and "TESTS:" in line:
```
Comment (philip-stoev) [inline misc/python/materialize/cli/mkrelease.py:325]: The splicing of the new section in mzcompose.yml appears to be incorrect:

```
  upgrade-from-latest:
    env:

  upgrade-from-0_8_4:
    env:
      PREVIOUS_VERSION: v0.8.3
      TESTS: any_version|any_version|v0.6.1|v0.7.3|v0.8.0|v0.8.1|v0.8.2|v0.8.3|v0.8.4
    steps:
      - step: workflow
        workflow: test-upgrade
      PREVIOUS_VERSION: latest
      TESTS: any_version|v0.6.1|v0.7.3|v0.8.0|v0.8.1|v0.8.2|v0.8.3|v0.8.4|latest_version
    steps:
      - step: workflow
        workflow: test-upgrade
```

So the  ```upgrade-from-0_8_4``` section landed in the middle of the ```upgrade-from-latest``` section.

To make life easier, you may wish to:
1. rearrange mzcompose.yml so that all the ``upgrade-from``` sections are at the very bottom
2. put a comment at the very end of the file that the items below that comment are generated automatically by mkrelease.py
3. Have mkrelease.py append the new section at the very end of the file rather than trying to splice it in.

**Review by philip-stoev — DISMISSED** [review]
I think the code that modifies mzcompose.yml is very fragile and has the potential of getting confused if the lines it expects are not at the relative offsets that it expects. Could you consider adding more assertions so that the entire thing fails if it is unable to lock on to any of the pieces of information that it expects?

I think having a YAML parser may be the safest solution, but it will require more hacking.

File: misc/python/materialize/cli/mkrelease.py (hunk 2 above)
Comment (quodlibetor) [reply misc/python/materialize/cli/mkrelease.py:319]: ah right.

File: misc/python/materialize/cli/mkrelease.py (hunk 1 above)
Comment (quodlibetor) [reply misc/python/materialize/cli/mkrelease.py:318]: ah, oops.

File: misc/python/materialize/cli/mkrelease.py (hunk 3 above)
Comment (quodlibetor) [reply misc/python/materialize/cli/mkrelease.py:325]: That's fair. I've now modified the script to correctly use the the "insert_workflow" variable that should be more robust to slight changes.

**Comment by quodlibetor** [comment]
> Could you consider adding more assertions so that the entire thing fails if it is unable to lock on to any of the pieces of information that it expects?

Yeah definitely I'll add some more. I'm also kind of leaning on the fact that this is test code and the tests themselves should fail if it's broken.

> I think having a YAML parser may be the safest solution, but it will require more hacking.

Yeah a yaml parser would probably be better for the workflow step because it is in a list and we can put the new step into the correct place. But unfortunately for the workflow definition there is no way to insert the new workflow in the "correct" place. If we don't care about the yaml code being well-organized, or this turns out to be fragile enough that we care less about the yaml being well-organized, then we should definitely switch to that.

**Comment by quodlibetor** [comment]
with my new version of the code and running it with "0.9.0" I get the following diff:

```diff
modified   test/upgrade/mzcompose.yml
@@ -30,6 +30,9 @@ mzworkflows:
       - step: workflow
         workflow: upgrade-from-0_8_3
 
+      - step: workflow
+        workflow: upgrade-from-0_9_0
+
       - step: workflow
         workflow: upgrade-from-latest
 
@@ -81,10 +84,18 @@ mzworkflows:
       - step: workflow
         workflow: test-upgrade
 
+  upgrade-from-0_9_0:
+    env:
+      PREVIOUS_VERSION: v0.9.0
+      TESTS: any_version|v0.6.1|v0.7.3|v0.8.0|v0.8.1|v0.8.2|v0.8.3|v0.9.0
+    steps:
+      - step: workflow
+        workflow: test-upgrade
+
   upgrade-from-latest:
     env:
       PREVIOUS_VERSION: latest
-      TESTS: any_version|v0.6.1|v0.7.3|v0.8.0|v0.8.1|v0.8.2|v0.8.3|latest_version
+      TESTS: any_version|v0.6.1|v0.7.3|v0.8.0|v0.8.1|v0.8.2|v0.8.3|v0.9.0|latest_version
     steps:
```

**Review by philip-stoev — APPROVED** [review]

**Review by philip-stoev — APPROVED** [review]
Thank you for the extra checks. The contents of the mzcompose.yml file appear to be as expected.

**Comment by benesch** [comment]
We are accreting a lot of tech debt here because we haven't addressed https://github.com/MaterializeInc/materialize/issues/6439. If the upgrade framework could be written in Python rather than as mzcompose files, then we wouldn't need this sort of brittle YAML-munging. I suppose this is the lesser evil right now, but .. I'd really like us to be wary of the amount of complexity that is creeping into mzrelease.

**Comment by quodlibetor** [comment]
> We are accreting a lot of tech debt here because we haven't addressed #6439.

Another option for this specifically would be to add custom `step` definitions for the tasks involved in the upgrade-test framework, or even something more dramatic like just moving upgrade-test into python that depends on the mzcompose code, without building out an entire `mzcompose.py` framework. I believe that would be equal or more code than is in this PR, though.

> I'd really like us to be wary of the amount of complexity that is creeping into mzrelease.

AFAICT The options are either for mzrelease to grow more complex or for the release process itself to grow more complex. Even if the code that we needed to add here was just checking if we need to append a version number to a text file, there would be _something_ going in to this. That is, if we want the entire release process to take less, not more, time.
